oauthenticator/google.py

```diff
@@ -127,10 +167,112 @@ def _cast_hosted_domain(self, proposal):
                 # unambiguous domain, use only base name
                 username = user_email.split('@')[0]
 
-        return {
-            'name': username,
-            'auth_state': {'access_token': access_token, 'google_user': bodyjs},
-        }
+        if self.admin_google_groups or self.google_group_whitelist:
+            credentials = await self._service_client_credentials(
+                    scopes=['https://www.googleapis.com/auth/admin.directory.group.readonly'],
+                    user_email_domain=user_email_domain)
+            groups_user_is_member_of = await self._google_groups_for_user(
+                    user_email=user_email,
+                    credentials=credentials)
+            bodyjs['google_groups'] = groups_user_is_member_of
+
+        # Check if user is a member of any admin groups.
+        is_admin = False
+        is_admin_group_specified = False
+        if self.admin_google_groups:
+            is_admin_group_specified = True
+            is_admin = await check_user_in_groups(member_groups=groups_user_is_member_of, allowed_groups=self.admin_google_groups[user_email_domain])
+
+        # Check if user is a member of any whitelisted groups.
+        user_in_group = False
+        is_group_specified = False
+        if self.google_group_whitelist:
+            is_group_specified = True
+            user_in_group = await check_user_in_groups(member_groups=groups_user_is_member_of, allowed_groups=self.google_group_whitelist[user_email_domain])
+
+        no_config_specified = not is_group_specified
```

Comment:
It took me some time to understand the conditionals here and I'm still not sure I got them right :(
But I believe we can simplifly it something like this:

```python
    if self.admin_google_groups or self.google_group_whitelist:
            credentials = await self._service_client_credentials(
                    scopes=['https://www.googleapis.com/auth/admin.directory.group.readonly'],
                    user_email_domain=user_email_domain)
            groups_user_is_member_of = await self._google_groups_for_user(
                    user_email=user_email,
                    credentials=credentials)
            bodyjs['google_groups'] = groups_user_is_member_of

            # Check if user is a member of any admin groups.
            is_admin = await check_user_in_groups(groups_user_is_member_of, self.admin_google_groups[user_email_domain])
            # Check if user is a member of any whitelisted groups.
            user_in_group = await check_user_in_groups(groups_user_is_member_of, self.google_group_whitelist[user_email_domain])

            if self.admin_google_groups:
                    return {
                        'name': username,
                        'auth_state': {'access_token': access_token, 'google_user': bodyjs},
                        'admin': is_admin,
                    }
            elif self.google_group_whitelist:
                if user_in_group:
                    return {
                        'name': username,
                        'auth_state': {'access_token': access_token, 'google_user': bodyjs},
                    }
                return None

    return {
        'name': username,
        'auth_state': {'access_token': access_token, 'google_user': bodyjs},
    }
```

What do you think @missingcharacter? Do you think this refactoring will work?

Reply:
Thanks, this works and it does look waaaay better